storage/pc-ata/ata.c

```diff
@@ -0,0 +1,533 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Generic ATA controller driver
+ *
+ * Copyright 2012-2015, 2019, 2020 Phoenix Systems
+ * Author: Marcin Stragowski, Kamil Amanowicz, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/list.h>
+#include <sys/threads.h>
+#include <sys/types.h>
+
+#include <arch/ia32/io.h>
+
+#include "ata.h"
+
+
+ata_common_t ata_common;
+
+
+static ata_bus_t *buses;
+
+
+static uint32_t ata_readreg(void *base, uint8_t reg, uint8_t size)
+{
+	uintptr_t reg_base = (uintptr_t)base;
+
+	if (reg_base & 0x1) {
+		/* Read from IO-port */
+		reg_base &= ~0x3;
+		reg_base += reg;
+
+		switch (size) {
+		case 1:
+			return inb((void *)(reg_base));
+		case 2:
+			return inw((void *)(reg_base));
+		case 4:
+			return inl((void *)(reg_base));
+		}
+	}
+	else {
+		/* Read from memory */
+		reg_base &= ~0xF;
+		reg_base += reg;
+
+		switch (size) {
+		case 1:
+			return *((uint8_t *)(reg_base));
+		case 2:
+			return *((uint16_t *)(reg_base));
+		case 4:
+			return *((uint32_t *)(reg_base));
+		}
+	}
+
+	return -1;
+}
+
+
+static void ata_writereg(void *base, uint8_t reg, uint32_t val, uint8_t size)
+{
+	uintptr_t reg_base = (uintptr_t)(base);
+
+	if (reg_base & 0x1) {
+		/* Write to IO-port */
+		reg_base &= ~0x3;
+		reg_base += reg;
+
+		switch (size) {
+		case 1:
+			outb((void *)(reg_base), (uint8_t)val);
+			break;
+		case 2:
+			outw((void *)(reg_base), (uint16_t)val);
+			break;
+		case 4:
+			outl((void *)(reg_base), val);
+			break;
+		}
+	}
+	else {
+		/* Write to memory */
+		reg_base &= ~0xF;
+		reg_base += reg;
+
+		switch (size) {
+		case 1:
+			*((uint8_t  *)(reg_base)) = (uint8_t)val;
+			break;
+		case 2:
+			*((uint16_t *)(reg_base)) = (uint16_t)val;
+			break;
+		case 4:
+			*((uint32_t *)(reg_base)) = val;
+			break;
+		}
+	};
+}
+
+
+static void ata_delay(ata_bus_t *bus)
+{
+	void *ctrl = bus->ctrl;
+
+	/* Single status/altstatus register read takes 100ns */
+	ata_readreg(ctrl, REG_ALTSTATUS, 1);
+	ata_readreg(ctrl, REG_ALTSTATUS, 1);
+	ata_readreg(ctrl, REG_ALTSTATUS, 1);
+	ata_readreg(ctrl, REG_ALTSTATUS, 1);
+}
+
+
+static void ata_resetbus(ata_bus_t *bus)
+{
+	void *ctrl = bus->ctrl;
+
+	/* Software reset all drives on the bus */
+	ata_writereg(ctrl, REG_CTRL, CTRL_SOFTRST | CTRL_NIEN, 1);
+	/* Short delay */

```

Comment:
This comment is rather redundant ;) I would rather know why we are waiting.

Reply:
Fixed.